src/StateManager.js

```diff
@@ -0,0 +1,64 @@
+// @flow
```

Comment:
What about extract the URL handling to a file of its own, and let the responsibility of the `StateManager` class to be delegating based on `mode`?

The directory structure could be something like:

```
├── src
│   ├── statemangement
│   │   ├── index.js // contains the delegator class (`StateManager`)
│   │   └── url-state-manager.js
```

Doing so separate concerns, which make the code easier to read, test, maintain and extend.

Reply:
Something like cf34821?